apollo-router/src/plugins/telemetry/apollo_otlp_exporter.rs

```diff
@@ -0,0 +1,323 @@
+            instrumentation_lib: self.intrumentation_library.clone(),
+        }
+    }
+
+    /// Adds the "apollo_private.operation.subtype" attribute for subscription requests.
+    fn prepare_execution_span(&self, mut span: LightSpanData) -> SpanData {
+        let op_type = span
+            .attributes
+            .get(&GRAPHQL_OPERATION_TYPE)
+            .and_then(extract_string)
+            .unwrap_or_default();
+        if op_type == OperationKind::Subscription.as_apollo_operation_type() {
+            // Currently, all "subscription" operations are of the "request" variety.
+            span.attributes.insert(KeyValue::new(
+                OPERATION_SUBTYPE,
+                OperationSubType::SubscriptionRequest.as_str(),
+            ));
+        }
+        self.base_prepare_span(span)
+    }
+
+    /// Parses and redacts errors from ftv1 traces.
+    /// Sets the span status to error if there are any errors.
+    fn prepare_subgraph_span(&self, mut span: LightSpanData) -> SpanData {
+        let mut status = Status::Unset;
+
+        // If there is an FTV1 attribute, process it for error redaction and replace it
+        if let Some(ftv1) = span.attributes.get(&APOLLO_PRIVATE_FTV1) {
+            let subgraph_name = span
+                .attributes
+                .get(&SUBGRAPH_NAME)
+                .and_then(extract_string)
+                .unwrap_or_default();
+            let subgraph_error_config = self
+                .errors_configuration
+                .subgraph
+                .get_error_config(&subgraph_name);
+            if let Some(Ok((trace_result, error_count))) =
+                extract_ftv1_trace_with_error_count(ftv1, subgraph_error_config)
+            {
+                if error_count > 0 {
+                    status = Status::error("ftv1")
+                }
+                let encoded = encode_ftv1_trace(&trace_result);
+                span.attributes
+                    .insert(KeyValue::new(APOLLO_PRIVATE_FTV1, encoded));
+            }
+        }
+
+        SpanData {
+            span_context: SpanContext::new(
+                span.trace_id,
+                span.span_id,
+                TraceFlags::default().with_sampled(true),
+                true,
+                TraceState::default(),
+            ),
+            parent_span_id: span.parent_span_id,
+            span_kind: span.span_kind.clone(),
+            name: span.name.clone(),
+            start_time: span.start_time,
+            end_time: span.end_time,
+            attributes: span.attributes,
+            events: EvictedQueue::new(0),
+            links: EvictedQueue::new(0),
+            status,
+            resource: Cow::Owned(self.resource_template.to_owned()),
+            instrumentation_lib: self.intrumentation_library.clone(),
+        }
+    }
+
+    pub(crate) fn export(&self, spans: Vec<SpanData>) -> BoxFuture<'static, ExportResult> {
+        let mut exporter = self.otlp_exporter.lock();
```

Comment:
This function makes me nervous. I tried converting it to be a straightforward `async` function and it wouldn't compile when I did that because it complained that the exporter lock wasn't `Send`.

I'm concerned that we take the lock before we pin the future along with the fact that the lock is held throughout the lifetime of the future.

Maybe take another look at this and see if we take the lock inside the future?

Reply:
I think that this is probablyt OK, as the exporter guard isn't moved inside a future, it's just locked to generate the future.

The code can be rewritten to make it more explicit:
```
let mut exporter = self.otlp_exporter.lock();
        let fut = exporter.export(spans);
        drop(exporter);
        Box::pin(fut.and_then(|_| {
```

Reply:
I like the suggestion from @BrynCooke to rewrite this more explicitly 👍

Reply:
https://github.com/apollographql/router/pull/4982/commits/c99082be3de4d3db1d4f03d227991e8e409dbdf8